### PR TITLE
feat(minecraft-extras): ComponentParser

### PR DESCRIPTION
**Pull request by Machine-Maker**
Adds a parser for adventure components using a provided ComponentDecoder. ~~Requries an adventure version bump to 4.16 for the ComponentDecoder interface.~~reverted

Currently I have the default StringMode to QUOTED for the static factory methods that don't take a StringMode parameter. Unsure if that's the best, but I don't think SINGLE is the best option as these can regularly be expected to contain spaces.

Remade from https://github.com/Incendo/cloud/pull/447

### CODE REVIEW

**Comment by Machine-Maker** [comment]
Implemented the brig mapping contributor mapping the ComponentParser to the brig string type based on the StringMode provided.

Also reverted the adventure version bump to not break compat.

File: cloud-brigadier/build.gradle.kts [hunk 1]
```diff
@@ -6,6 +6,6 @@ plugins {
 dependencies {
     api(libs.cloud.core)
     /* Needs to be provided by the platform */
-    compileOnly(libs.brigadier)
+    compileOnlyApi(libs.brigadier)
```
Comment (jpenilla) [inline cloud-brigadier/build.gradle.kts:9]: this can't be exposed as it's not on central

File: cloud-minecraft-extras/build.gradle.kts [hunk 2]
```diff
@@ -5,6 +5,7 @@ plugins {
 
 dependencies {
     api(libs.cloud.core)
+    api(projects.cloudBrigadier)
```
Comment (jpenilla) [inline cloud-minecraft-extras/build.gradle.kts:8]: compileOnly this and brigadier

**Comment by jpenilla** [comment]
Still needs a ParserContributor, not sure how we want to handle selecting a serializer from annotations though. It should also check the string greedy/quoted annotations.

**Comment by Machine-Maker** [comment]
Ok, add a ParserContributor and annotations to configure the ComponentParser

File: cloud-minecraft-extras/src/main/java/org/incendo/cloud/minecraft/extras/MinecraftExtraParserParameters.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import org.apiguardian.api.API;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.incendo.cloud.parser.ParserParameter;
+
+/**
+ * {@link ParserParameter} keys for cloud-minecraft-extras
+ * @since 2.0.0
+ */
+@API(status = API.Status.STABLE, since = "2.0.0")
+public final class MinecraftExtraParserParameters {
```
Comment (jpenilla) [inline cloud-minecraft-extras/src/main/java/org/incendo/cloud/minecraft/extras/MinecraftExtraParserParameters.java:38]:
```suggestion
public final class MinecraftExtrasParserParameters {
```

File: cloud-minecraft-extras/src/main/java/org/incendo/cloud/minecraft/extras/parser/ComponentParserContributor.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                    stringMode = StringParser.StringMode.QUOTED;
+                } else {
+                    stringMode = StringParser.StringMode.SINGLE;
+                }
+                final Function<String, ? extends Component> decoder =
+                    options.get(MinecraftExtraParserParameters.COMPONENT_DECODER, MiniMessage.miniMessage()::deserialize);
+                return new ComponentParser<>(decoder, stringMode);
+            });
+        } catch (final Exception | LinkageError ignore) {
+            // ignore
```
Comment (jpenilla) [inline cloud-minecraft-extras/src/main/java/org/incendo/cloud/minecraft/extras/parser/ComponentParserContributor.java:103]: when would this happen? the lambdas shouldn't run unless the user uses the annotation

Comment (Machine-Maker) [reply cloud-minecraft-extras/src/main/java/org/incendo/cloud/minecraft/extras/parser/ComponentParserContributor.java:103]: I was just following the pattern of the text color parser contributor

Comment (jpenilla) [reply cloud-minecraft-extras/src/main/java/org/incendo/cloud/minecraft/extras/parser/ComponentParserContributor.java:103]: Oh I remember, it's for when extras happens to be on the classpath but adventure isn't... should have left a better comment

File: gradle/build-logic/src/main/kotlin/conventions.base.gradle.kts [hunk 5]
```diff
@@ -4,6 +4,7 @@ plugins {
     id("org.incendo.cloud-build-logic")
     id("org.incendo.cloud-build-logic.spotless")
     id("org.incendo.cloud-build-logic.errorprone")
+    id("org.incendo.cloud-build-logic.javadoc-links")
```
Comment (jpenilla) [inline gradle/build-logic/src/main/kotlin/conventions.base.gradle.kts:7]: why was this added here?

File: cloud-minecraft-extras/build.gradle.kts [hunk 6]
```diff
@@ -17,6 +19,8 @@ dependencies {
     }
     compileOnlyApiAndTests(libs.adventureApi)
     compileOnlyApiAndTests(libs.adventureTextSerializerPlain)
+    compileOnlyApiAndTests(libs.adventureTextSerializerLegacy)
```
Comment (jpenilla) [inline cloud-minecraft-extras/build.gradle.kts:22]: These should just be compileOnly, not compileOnlyApi. Should also add a comment that these are for ComponentArgument annotation mapping

File: gradle/build-logic/src/main/kotlin/conventions.base.gradle.kts (hunk 5 above)
Comment (Machine-Maker) [reply gradle/build-logic/src/main/kotlin/conventions.base.gradle.kts:7]: oh, forgot to remove it. leftover from when I had to exclude brig from javadocs when I was including brig as an api dependency

File: cloud-minecraft-extras/src/main/java/org/incendo/cloud/minecraft/extras/annotation/Decoder.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//
+package org.incendo.cloud.minecraft.extras.annotation;
```
Comment (jpenilla) [inline cloud-minecraft-extras/src/main/java/org/incendo/cloud/minecraft/extras/annotation/Decoder.java:24]:
```suggestion
package org.incendo.cloud.minecraft.extras.annotation.specifier;
```

Comment (Machine-Maker) [reply cloud-minecraft-extras/src/main/java/org/incendo/cloud/minecraft/extras/annotation/Decoder.java]: done

File: examples/example-bukkit/src/main/java/org/incendo/cloud/examples/bukkit/builder/feature/minecraft/ComponentExample.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    @Override
+    public void registerFeature(
+        final @NonNull ExamplePlugin examplePlugin,
+        final @NonNull BukkitCommandManager<CommandSender> manager
+    ) {
+        manager.command(
+            manager.commandBuilder("builder")
+                .literal("minimessage")
+                .required("msg", ComponentParser.componentParser(
+                    MiniMessage.miniMessage()::deserialize,
```
Comment (jpenilla) [inline examples/example-bukkit/src/main/java/org/incendo/cloud/examples/bukkit/builder/feature/minecraft/ComponentExample.java:49]:
```suggestion
                    MiniMessage.miniMessage(),
```
this should work too, right?

Comment (jpenilla) [reply examples/example-bukkit/src/main/java/org/incendo/cloud/examples/bukkit/builder/feature/minecraft/ComponentExample.java:49]: Or even the miniMessageParser shorthand... I guess this is fine for the example to show using a generic function.

**Review by jpenilla — APPROVED** [review]

File: examples/example-bukkit/src/main/java/org/incendo/cloud/examples/bukkit/builder/feature/minecraft/ComponentExample.java (hunk 8 above)
Comment (Machine-Maker) [reply examples/example-bukkit/src/main/java/org/incendo/cloud/examples/bukkit/builder/feature/minecraft/ComponentExample.java:49]: yeah. How many examples usages do you think I should add there?
